plugins/synaptics-rmi/fu-synaptics-rmi-v5-device.c

```diff
@@ -0,0 +1,269 @@
+		return FALSE;
+
+	/* all other versions */
+	fu_byte_array_append_uint8 (erase_cmd, RMI_F34_ERASE_ALL);
+	if (!fu_synaptics_rmi_device_write (self,
+					    flash->status_addr,
+					    erase_cmd,
+					    error)) {
+		g_prefix_error (error, "failed to erase core config: ");
+		return FALSE;
+	}
+	g_usleep (1000 * RMI_F34_ENABLE_WAIT_MS);
+	if (!fu_synaptics_rmi_device_wait_for_idle (self,
+						    RMI_F34_ERASE_WAIT_MS,
+						    RMI_DEVICE_WAIT_FOR_IDLE_FLAG_REFRESH_F34,
+						    error)) {
+		g_prefix_error (error, "failed to wait for idle for erase: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_synaptics_rmi_v5_device_write_block (FuSynapticsRmiDevice *self,
+					guint8 cmd,
+					guint32 address,
+					const guint8 *data,
+					gsize datasz,
+					GError **error)
+{
+	g_autoptr(GByteArray) req = g_byte_array_new ();
+
+	g_byte_array_append (req, data, datasz);
+	fu_byte_array_append_uint8 (req, cmd);
+	if (!fu_synaptics_rmi_device_write (self, address, req, error)) {
+		g_prefix_error (error, "failed to write block @0x%x: ", address);
+		return FALSE;
+	}
+	if (!fu_synaptics_rmi_device_wait_for_idle (self,
+						    RMI_F34_IDLE_WAIT_MS,
+						    RMI_DEVICE_WAIT_FOR_IDLE_FLAG_NONE,
+						    error)) {
+		g_prefix_error (error, "failed to wait for idle @0x%x: ", address);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+gboolean
+fu_synaptics_rmi_v5_device_write_firmware (FuDevice *device,
+					   FuFirmware *firmware,
+					   FwupdInstallFlags flags,
+					   GError **error)
+{
+
+	FuSynapticsRmiDevice *self = FU_SYNAPTICS_RMI_DEVICE (device);
+	FuSynapticsRmiFlash *flash = fu_synaptics_rmi_device_get_flash (self);
+	FuSynapticsRmiFunction *f34;
+	guint32 address;
+	g_autoptr(GBytes) bytes_bin = NULL;
+	g_autoptr(GBytes) bytes_cfg = NULL;
+	g_autoptr(GPtrArray) chunks_bin = NULL;
+	g_autoptr(GPtrArray) chunks_cfg = NULL;
+	g_autoptr(GByteArray) req_addr = g_byte_array_new ();
+
+	/* we should be in bootloader mode now, but check anyway */
+	if (!fu_device_has_flag (device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER)) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "not bootloader, perhaps need detach?!");
+		return FALSE;
+	}
+
+	/* check is idle */
+	if (!fu_synaptics_rmi_device_wait_for_idle (self, 0,
+						    RMI_DEVICE_WAIT_FOR_IDLE_FLAG_REFRESH_F34,
+						    error)) {
+		g_prefix_error (error, "not idle: ");
+		return FALSE;
+	}
+
+	/* f34 */
+	f34 = fu_synaptics_rmi_device_get_function (self, 0x34, error);
+	if (f34 == NULL)
+		return FALSE;
+
+	/* get both images */
+	bytes_bin = fu_firmware_get_image_by_id_bytes (firmware, "ui", error);
+	if (bytes_bin == NULL)
+		return FALSE;
+	bytes_cfg = fu_firmware_get_image_by_id_bytes (firmware, "config", error);
+	if (bytes_cfg == NULL)
+		return FALSE;
+
+	/* disable powersaving */
+	if (!fu_synaptics_rmi_device_disable_sleep (self, error)) {
+		g_prefix_error (error, "failed to disable sleep: ");
```

Comment:
this prefix is only in v5, not v7.  intentional?

Reply:
I *think* so -- v7 might need a different sleep function. If not I can unexport it i guess.

Reply:
Ohh, we do need this for v7.